### ISSUE TITLE
Reconciling properties

### ISSUE THREAD

**Issue by eudemocracystiftung** [OpenRefine/OpenRefine#6842]
Our project has a lot of financial data, including hundreds of audit reports which we are trying to enter into a Wikibase. The table of audit data has the following columns:

*Entity
*Financial year
*Audit field (e.g. "salaries", "meeting costs", "admin expenditure", etc.)
*Amount

The table is over 8,000 lines long. We can easily reconciled the first column, as entities exist as items.

However, we are unable to reconcile the third column, as audit fields are properties -- all of type quantity, with the fourth column (amount) as the value.

Under the current system, we need to import the audit fields one by one, but we have over a hundred of them.

### Proposed solution

As discussed on the [OpenRefine forum](https://forum.openrefine.org/t/reconciling-properties/1744), the solution would be to allow for the reconciliation not just of items, but also of properties.

Quoting antonin_d:

> The first step would obviously be to use a reconciliation service that can reconcile properties. The default Wikidata service cannot and I don't know any that can as of today, but it would likely not be very hard to build.

> Assuming we have that, it doesn't solve your problem entirely though. You'd need to be able to drag and drop reconciled properties in a schema, and that's not supported yet. It's not so clear to me how this could be built, because when adding a statement (or qualifier, or reference part) we need to know the property's datatype to be able to display an appropriate field for the value.
In general, your column of reconciled properties could potentially contain properties with different datatypes! I can think of a few approaches:

> - build the reconciliation service so that you can reconcile to properties of a specific datatype, ensuring all reconciled values are of a uniform datatype, which can then be recognized by the schema editor to offer the appropriate input field for the value.
> - somehow let the user select themselves which datatype should be used in the schema editor. Find a way to report errors appropriately when the datatype of the property and of the value don't match.
> - accept mixed datatypes in the column of reconciled property, and offer a generic input field for the value, which intelligently parses whatever is input there to match the datatype of the property. It's a bit cumbersome to build and the UX might be a bit confusing because the user needs to know how to format the values so that they are understood correctly. For datatypes like quantities, where we currently display an amount and a unit field, we'd need to offer a different syntax to specify the amount and unit in a single text field, perhaps. And good error reporting would also be needed.

EDIT: changed "meetings" with "meeting costs" for clarity.

**Comment by thadguidry** [comment]
Are you sure the 3rd column "audit field" are all type quantity?  "meetings" has a quantity?
So it would be something like
|Entity|Financial Year|Audit Field|Quantity Amount|
|----|----|----|----|
|Information Systems Cooperation Group|2023|Meetings|3|

`meetings = (qty) 3`
"they had 3 meetings about this entity" ???

**Comment by thadguidry** [comment]
makes sense.  But I do struggle with this question in my mind "why would they want to use a graph database like Wikidata for financial audit data, than say, a RDBMS or even a Document database?  I guess they want to see relationships with their entities to other things and not only relating audit data between entities?"

**Comment by tfmorris** [comment]
This feels more like a data modeling issue than a tool functionality issue. The "audit field," based on the examples is actually something like an "expense category" or "balance sheet account" or something along those lines. That would make "Personnel Costs" an instance, not a property.

**Comment by wetneb** [comment]
@tfmorris are you then suggesting that they use a single, generic property with quantity datatype, whose meaning would be determined by another property added as a qualifier (say, item-valued)? It might be a sensible choice in some cases but it also feels legitimate to me to use specific properties, as it gives a much cleaner statement structure (which also exports better to RDF, for instance).

In general, when users upload data to a Wikibase, they are often not in control of the ontology they must use. Typically, in Wikidata, the ontology is curated by the community and property creation is guarded by an approval process. So in my opinion the need that is voiced here is a legitimate one (albeit not occurring very often), that OpenRefine should ideally support.

**Comment by wetneb** [comment]
I wouldn't expect much progress on this soon, on the OpenRefine side of things, because it hasn't been identified as high priority for Wikimedia Commons (the current improvements around Wikibase integration are focused on this platform).

Work on making a reconciliation service for properties could start independently of that. I am part of an effort over at the [Wikibase Stakeholder Group (WBSG)](https://wbstakeholder.group/) to develop a new reconciliation service for Wikibase, which would be directly served by Wikibase itself (without the need to run an external Python wrapper). Adding support for properties in this new reconciliation service has been mentioned as a potential stretch goal. We are currently trying to fundraise from various source to get sufficient funding to commission the development of this MediaWiki extension. If you have ideas of institutions which could contribute to this effort, let me know! And perhaps you could be interested in joining the WBSG (if you indeed arren't represented there yet)?
